### PR TITLE
Comms Connection

### PR DESCRIPTION
**Pull request by GordonSmith**
Signed-off-by: Gordon Smith <gordonjsmith@gmail.com>

### CODE REVIEW

File: src/util/logging.ts [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        return this;
+    }
+
+    pushLevel(_: Level): this {
+        this._levelStack.push(this._level);
+        this._level = _;
+        return this;
+    }
+
+    popLevel(_: Level): this {
```
Comment (GordonSmith) [inline src/util/logging.ts]: Remove param

**Review by GordonSmith — COMMENTED** [review]
Add unit tests for util functions

File: src/comms/connection.ts [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+            const xhr = _d3Request(options.uri);
+            if (verb === "POST") {
+                xhr
+                    .header("X-Requested-With", "XMLHttpRequest")
+                    .header("Content-Type", "application/x-www-form-urlencoded")
+                    .header("Origin", null)
+                    ;
+            }
+            xhr
+                .send(verb, options.body, (err: any, req: any) => {
```
Comment (GordonSmith) [inline src/comms/connection.ts]: Rename to response

**Comment by GordonSmith** [comment]
@miguelvazq @jbrundage Updated with missing unit tests (stack, string + url) also fixed two minor issues I commented on during the group review.

**Comment by jbrundage** [comment]
Maybe the *src/util/* directory should be abstracted into its own *hpcc-js-util* repository? I could see these utility files being useful in hpcc-js-viz as well.

**Review by jbrundage — APPROVED** [review]
Conceptually everything looks fine to me.

**Comment by jbrundage** [comment]
The improved error handling is going to be great. It would be an interesting addition to use localStorage to keep a running log of request times.

**Comment by GordonSmith** [comment]
@jbrundage  re the "util" being its own package - that has already crossed my mind.  (at least now with es6 / ts proper modules, refactoring and moving these things around is a trivial exercise).

File: src/comms/connection.ts [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+                case "POST":
+                    options.headers = {
+                        "X-Requested-With": "XMLHttpRequest",
+                        "Content-Type": "application/x-www-form-urlencoded"
+                    };
+                    options.rejectUnauthorized = this._opts.rejectUnauthorized;
+                    options.body = this.serialize(request);
+                    break;
+                default:
+            }
```
Comment (miguelvazq) [inline src/comms/connection.ts]: Minor: Missing ;

File: src/comms/connection.ts [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+                    clearInterval(progress);
+                } else {
+                    respondedTimeout -= respondedTick;
+                    if (respondedTimeout <= 0) {
+                        clearInterval(progress);
+                        // logger.debug("Request timeout:  " + script.src);
+                        doCallback();
+                        reject(Error("Request timeout:  " + script.src));
+                    } else {
+                        // logger.debug("Request pending (" + respondedTimeout / 1000 + " sec):  " + script.src);
```
Comment (miguelvazq) [inline src/comms/connection.ts]: else does nothing

**Review by miguelvazq — COMMENTED** [review]
Just one small comment

File: src/comms/connection.ts (hunk 4 above)
Comment (GordonSmith) [reply src/comms/connection.ts]: @miguelvazq Change made (added logging back in)

**Comment by miguelvazq** [comment]
Looks good to me.

**Pull request merged**